pkg/resource/deploy/step_executor.go

```diff
@@ -0,0 +1,305 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/util/cancel"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+)
+
+const (
+	// Dummy workerID for synchronous operations.
+	synchronousWorkerID = -1
+
+	// Utility constant for easy debugging.
+	stepExecutorLogLevel = 4
+)
+
+// A Chain is a sequence of Steps that must be executed in the given order.
+type Chain = []Step
+
+// stepExecutor is the component of the engine responsible for taking steps and executing
+// them, possibly in parallel if requested. The step generator operates on the granularity
+// of "chains", which are sequences of steps that must be executed exactly in the given order.
+// Chains are a simplification of the full dependency graph DAG within Pulumi programs. Since
+// Pulumi language hosts can only invoke the resource monitor once all of their dependencies have
+// resolved, we (the engine) can assume that any chain given to us by the step generator is already
+// ready to execute.
+type stepExecutor struct {
+	plan        *Plan    // The plan currently being executed.
+	opts        Options  // The options for this current plan.
+	preview     bool     // Whether or not we are doing a preview.
+	pendingNews sync.Map // Resources that have been created but are pending a RegisterResourceOutputs.
+
+	workers        sync.WaitGroup // WaitGroup tracking the worker goroutines that are owned by this step executor.
+	incomingChains chan Chain     // Incoming chains that we are to execute
+
+	cancel *cancel.Source // Cancellation source for this step executor.
+
+	sawError uint32 // atomic boolean indicating whether or not the step excecutor saw that there was an error.
+}
+
+//
+// The stepExecutor communicates with a stepGenerator by listening to a channel. As the step generator
+// generates new chains that need to be executed, the step executor will listen to this channel to execute
+// those steps.
+//
+
+// Execute submits a Chain for asynchronous execution. The execution of the chain will begin as soon as there
+// is a worker available to execute it.
+func (se *stepExecutor) Execute(chain Chain) {
+	select {
+	case se.incomingChains <- chain:
+	case <-se.cancel.Context().Canceled():
+	}
+}
+
+// ExecuteRegisterResourceOutputs services a RegisterResourceOutputsEvent synchronously on the calling goroutine.
+func (se *stepExecutor) ExecuteRegisterResourceOutputs(e RegisterResourceOutputsEvent) {
+	// Look up the final state in the pending registration list.
+	urn := e.URN()
+	value, has := se.pendingNews.Load(urn)
+	reg := value.(Step)
+	contract.Assertf(has, "cannot complete a resource '%v' whose registration isn't pending", urn)
```

Comment:
Should the asserts come before the prior line, since `value` will be `nil` and the prior line would panic anyway?

Reply:
yep, you're right. This is not actually new code, so good catch 😄